### PR TITLE
Deregister event handlers when stopping server with event handler made auto-closeable

### PR DESCRIPTION
**Pull request by bowenliang123**
### _Why are the changes needed?_

- deregister event handlers when stopping Kyuubiserver, by deregister them from EventBus's handler Registry
- change `EventHandler` from `type` to `trait` and make it extending  `AutoCloseable`
- implement `close` method in `JsonLoggingEventHandler` for closing writers and streams

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala [hunk 1]
```diff
@@ -122,5 +133,10 @@ object EventBus extends Logging {
       } yield parent
       clazz :: parents
     }
+
+    def deregisterAll(): Unit = {
```
Comment (pan3793) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala:137]: synchronized

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala [hunk 2]
```diff
@@ -96,6 +102,11 @@ object EventBus extends Logging {
       asyncEventHandlerRegistry.register(et)
       this
     }
+
+    override def deregisterAll(): Unit = {
```
Comment (pan3793) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala:106]: synchronized

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala [hunk 3]
```diff
@@ -68,6 +70,10 @@ object EventBus extends Logging {
   def registerAsync[T <: KyuubiEvent: ClassTag](et: EventHandler[T]): EventBus =
     defaultEventBus.registerAsync[T](et)
 
+  def deregisterAll(): Unit = {
+    defaultEventBus.deregisterAll()
```
Comment (pan3793) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala:74]: synchronized

Comment (bowenliang123) [reply kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala:74]: Which one are you suggesting? Synchronized on the object of EventBus instance or on the defaultEventBus instance?

```
def deregisterAll(): Unit = synchronized {
```
or
```
def deregisterAll(): Unit = defaultEventBus.synchronized {
```

**Review by pan3793 — APPROVED** [review]

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/handler/JsonLoggingEventHandler.scala [hunk 4]
```diff
@@ -108,6 +119,7 @@ class JsonLoggingEventHandler(
 }
 
 object JsonLoggingEventHandler {
-  val JSON_LOG_DIR_PERM: FsPermission = new FsPermission(Integer.parseInt("770", 8).toShort)
-  val JSON_LOG_FILE_PERM: FsPermission = new FsPermission(Integer.parseInt("660", 8).toShort)
+  private val JSON_LOG_DIR_PERM: FsPermission = new FsPermission(Integer.parseInt("770", 8).toShort)
+  private val JSON_LOG_FILE_PERM: FsPermission =
```
Comment (yaooqinn) [inline kyuubi-events/src/main/scala/org/apache/kyuubi/events/handler/JsonLoggingEventHandler.scala:123]: why changing these？

**Comment by yaooqinn** [comment]
Can we add some tests?

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/handler/JsonLoggingEventHandler.scala (hunk 4 above)
Comment (bowenliang123) [reply kyuubi-events/src/main/scala/org/apache/kyuubi/events/handler/JsonLoggingEventHandler.scala:123]: They could be private only and the changes are optional. This pr is closing the inside intermediateness of the json event handlers, so I think good to retract these two variables in private accessibility.

File: kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala (hunk 3 above)
Comment (pan3793) [reply kyuubi-events/src/main/scala/org/apache/kyuubi/events/EventBus.scala:74]: the former

**Comment by bowenliang123** [comment]
> Can we add some tests?

I think the changes in `[ServerJsonLoggingEventHandlerSuite.scala](https://github.com/apache/kyuubi/pull/4777/files#diff-91972dfd64cc4b9aa688957a9b4e2a78f328a1cdc26beae3048a8a62fb9b9a10R201)` reflects the situations. It is equivalent to testing both no messages after clearing up the event handlers and receiving messages by re-registering event handlers again.

**Review by wForget — APPROVED** [review]

**Comment by bowenliang123** [comment]
Thanks, merged to master.

**Pull request closed without merging**